util/progress/progressui/json.go

```diff
@@ -0,0 +1,61 @@
+package progressui
+
+import (
+	"time"
+
+	"github.com/moby/buildkit/solver/pb"
+	digest "github.com/opencontainers/go-digest"
+)
+
+// SolveStatus holds the latest status update for a vertex by its digest.
+type SolveStatus struct {
+	Digest   digest.Digest   `json:"digest"`
+	Vertex   Vertex          `json:"vertex,omitempty"`
+	Statuses []VertexStatus  `json:"statuses,omitempty"`
+	Warnings []VertexWarning `json:"warnings,omitempty"`
+	Logs     []VertexLog     `json:"logs,omitempty"`
+}
+
+// Vertex contains information about the vertex itself.
+// Information will only be included if it has changed since the last time
+// that the information was sent or if it is the first time this vertex has
+// been output.
+type Vertex struct {
```

Comment:
I believe everything but `SolveStatus` is already identical to the defined types. In that case, combine the types or use type aliases.

edit: I see the `Vertex digest.Digest` is different that could be added with a wrapper, embedding the shared object. For the `VertexLog`, the already defined type with `Stream` and without `Partial` is better.

Reply:
I think it's probably better to just keep them separate so we aren't tying the two APIs together. Yes, they'll share a lot of common attributes, but they're still different APIs. There is a bit of a difference here as you noted. This API kind of guarantees that the update struct will only contain data from one vertex rather than potentially having multiple different vertex updates in the same batch.

Reply:
Otoh I don't want these to start to diverge. I think that is a bigger risk. We can move types to separate packages if the imports would become too big. This can be done later but mark it as follow-up after this PR merges then.

Reply:
Do we want to just expose the raw solve status struct as-is? That might be better.

Reply:
You mean the raw event structs as they appear from protobuf? In that case, it should be `jsonraw`. It's not as useful for users because the only way to understand it would be to convert it into structured form first (something that can't be done with jq for example).